scanner/src/main/kotlin/scanners/fossid/FossId.kt

```diff
@@ -0,0 +1,396 @@
+     */
+    private suspend fun wait(timeout: Long, delay: Long, waitLoop: suspend () -> Boolean) =
+        withTimeoutOrNull(timeout) {
+            var isFinished = false
+            while (!isFinished) {
+                isFinished = waitLoop()
+
+                if (!isFinished) {
+                    delay(delay)
+                }
+            }
+        }
+
+    /**
+     * Wait until the repository of a scan with [scanCode] has been downloaded.
+     */
+    private suspend fun waitDownloadComplete(scanCode: String) {
+        val timeout = wait(WAIT_INTERVAL_MS * WAIT_REPETITION, WAIT_INTERVAL_MS) {
+            this@FossId.log.info { "Check download status for '$scanCode'." }
+
+            val response = service.checkDownloadStatus(user, apiKey, scanCode)
+                .checkResponse("check download status")
+
+            if (response.data == DownloadStatus.FINISHED) {
+                true
+            } else {
+                /*
+        There is a weird bug with the FossId server with version < 20.2: sometimes the download is
+        complete but it stays in state "NOT FINISHED". Therefore we check the output of Git Fetch to
+        find out if the download is indeed done
+                 */
+                val message = response.message
+                if (message == null || !GIT_FETCH_DONE_PATTERN.matcher(message).find()) {
+                    false
+                } else {
+                    this@FossId.log.warn {
+                        "The download is not finished but Git Fetch has completed. Carrying on..."
+                    }
+                    true
+                }
+            }
+        }
+        requireNotNull(timeout) {
+            "Timeout while waiting for the download to complete"
+        }
+        log.info("Data download has been completed.")
+    }
+
+    /**
+     * Wait until a scan with [scanCode] has completed.
+     */
+    private suspend fun waitScanComplete(scanCode: String) {
+        val timeout = wait(WAIT_INTERVAL_MS * WAIT_REPETITION, WAIT_INTERVAL_MS) {
+            this@FossId.log.info { "Waiting for scan='$scanCode' to complete." }
+
+            val response = service.checkScanStatus(user, apiKey, scanCode)
+                .checkResponse("check scan status", false)
+
+            if (response.data == null) {
+                false
+            } else {
+                when (response.data?.state) {
+                    ScanState.FINISHED -> {
+                        this@FossId.log.info { "Scan result: ${response.data?.comment}." }
+                        true
+                    }
+                    else -> {
+                        this@FossId.log.info {
+                            "Scan status for '$scanCode' is '${response.data?.state}'. " +
+                                    "Waiting..."
+                        }
+                        false
+                    }
+                }
+            }
```

Comment:
Probably a bit more readable:

            response.data?.let {
                if (it.state == ScanState.FINISHED) {
                    FossId.log.info { "Scan result: ${response.data?.comment}." }
                    true
                } else {
                    FossId.log.info { "Scan status for '$scanCode' is '${response.data?.state}'. Waiting..." }
                    false
                }
            } ?: false